doc/man/readall.1
.\" $Id: readall.1,v 1.1 2002/02/01 21:07:05 engelhar Exp $
.TH readall 1 "2/1/2002" "" ""
.SH NAME
readall - read the current value of all Web100 variables from all connections.
.SH SYNOPSIS
.B readall
.SH DESCRIPTION
\fIreadall\fP iterates over all connections, grouping and printing out each
Web100 variable from each one.
.SH SEE ALSO
.BR util (1),
.BR readvar (1),
.BR writevar (1),
.BR deltavar (1),
.BR web100 (7)


NEWS
News
====

    This document gives an overview of the major changes of the software since
the last release.

  Alpha 1.2pre2:
    o Fixed the binary interface incompatibilities that were introduced
      in Alpha 1.2pre1.  This means that programs which were compiled
      against Alpha 1.1 should continue to work unchanged with the
      Alpha 1.2pre2 libweb100.

  Alpha 1.2pre1:
    o NOTE: The binary interface to libweb100 has changed slightly.  This
      means that existing applications must be recompiled and relinked
      with libweb100.  However, it is very likely that the recompile will
      not require changes to existing source code.
    o Added support for both IPv4 and IPv6 types from the kernel (previous
      versions of the library would crash if the kernel returned IPv6
      values).
    o Added support for logging values from the kernel
    o Addition of 'readvarm' and 'readconn' scripts
    o Addition of man pages for many functions, such as the snapshot and
      logging functions (some are only stubs).
    o Many GUI improvements and cleanups
    o Updates and cleanups to the developer's and user's guide.
    o Removed some items from web100.h that were not meant to be public
    o Support building on systems which do not have socklen_t

  Alpha 1.1:
    o GUI fixes: Renamed "util" to "gutil", fixed the vdt display,
      consolidation of code, miscellaneous bug and display fixes.
    o writevar fixed so that it works for integers
    o Added web100_connection_from_socket() and web100_value_to_textn()
      to libweb100.
    o Bug fixes in libweb100, such as for _find() and _lookup()
      functions.
    o Documentation updates and additions, such as the developer's
      guide, user's guide, and man pages for some libweb100 API
      functions (see "man libweb100").

  Alpha 1.0:
    o First public release


util/scripts/readall.c
/*
 * readall: read all variables from all connections and print them to stdout.
 *
 * Copyright (c) 2001
 *      Carnegie Mellon University, The Board of Trustees of the University
 *      of Illinois, and University Corporation for Atmospheric Research.
 *      All rights reserved.  This software comes with NO WARRANTY.
 *
 * Since our code is currently under active development we prefer that
 * everyone gets the it directly from us.  This will permit us to
 * collaborate with all of the users.  So for the time being, please refer
 * potential users to us instead of redistributing web100.
 *
 * $Id: readall.c,v 1.3 2002/09/05 20:21:29 jheffner Exp $
 */
#include <stdio.h>
#include <stdlib.h>

#include "web100.h"


int main(int argc, char *argv[])
{
    web100_agent *agent;
    web100_group *group;
    web100_group *read_grp;
    web100_var *addr_type, *laddr, *raddr, *lport, *rport;

    if ((agent = web100_attach(WEB100_AGENT_TYPE_LOCAL, NULL)) == NULL) {
        web100_perror("web100_attach");
        exit(EXIT_FAILURE);
    }

    if ((read_grp = web100_group_find(agent, "read")) == NULL) {
        web100_perror("web100_group_find: read");
        exit(EXIT_FAILURE);
    }
    if ((addr_type =
         web100_var_find(read_grp, "LocalAddressType")) == NULL) {
        web100_perror("web100_var_find: LocalAddressType");
        exit(EXIT_FAILURE);
    }
    if ((laddr = web100_var_find(read_grp, "LocalAddress")) == NULL) {
        web100_perror("web100_var_find: LocalAddress");
        exit(EXIT_FAILURE);
    }
    if ((raddr = web100_var_find(read_grp, "RemAddress")) == NULL) {
        web100_perror("web100_var_find: RemAddress");
        exit(EXIT_FAILURE);
    }
    if ((lport = web100_var_find(read_grp, "LocalPort")) == NULL) {
        web100_perror("web100_var_find: LocalPort");
        exit(EXIT_FAILURE);
    }
    if ((rport = web100_var_find(read_grp, "RemPort")) == NULL) {
        web100_perror("web100_var_find: RemPort");
        exit(EXIT_FAILURE);
    }

    group = web100_group_head(agent);

    while (group) {
        web100_connection *conn;

        printf("Group \"%s\"\n", web100_get_group_name(group));

        conn = web100_connection_head(agent);

        while (conn) {
            web100_var *var;
            web100_snapshot *snap;
            int type;
            int cid;
            char buf[256];

            cid = web100_get_connection_cid(conn);
            printf("Connection %d (", cid);

            if (web100_raw_read(addr_type, conn, buf) !=
                WEB100_ERR_SUCCESS) {
                web100_perror("web100_raw_read");
                exit(EXIT_FAILURE);
            }
            type = *(int *) buf;
            type =
                (type ==
                 WEB100_ADDRTYPE_IPV4 ? WEB100_TYPE_INET_ADDRESS_IPV4 :
                 WEB100_TYPE_INET_ADDRESS_IPV6);
            if (web100_raw_read(laddr, conn, buf) != WEB100_ERR_SUCCESS) {
                web100_perror("web100_raw_read");
                exit(EXIT_FAILURE);
            }
            printf("%s ", web100_value_to_text(type, buf));
            if (web100_raw_read(lport, conn, buf) != WEB100_ERR_SUCCESS) {
                web100_perror("web100_raw_read");
                exit(EXIT_FAILURE);
            }
            printf("%s  ",
                   web100_value_to_text(WEB100_TYPE_INET_PORT_NUMBER,
                                        buf));
            if (web100_raw_read(raddr, conn, buf) != WEB100_ERR_SUCCESS) {
                web100_perror("web100_raw_read");
                exit(EXIT_FAILURE);
            }
            printf("%s ", web100_value_to_text(type, buf));
            if (web100_raw_read(rport, conn, buf) != WEB100_ERR_SUCCESS) {
                web100_perror("web100_raw_read");
                exit(EXIT_FAILURE);
            }
            printf("%s)\n",
                   web100_value_to_text(WEB100_TYPE_INET_PORT_NUMBER,
                                        buf));

            if ((snap = web100_snapshot_alloc(group, conn)) == NULL) {
                web100_perror("web100_snapshot_alloc");
                exit(EXIT_FAILURE);
            }

            if (web100_snap(snap)) {
                perror("web100_snap");
                if (web100_errno == WEB100_ERR_NOCONNECTION)
                    continue;
                exit(EXIT_FAILURE);
            }

            var = web100_var_head(group);

            while (var) {
                if (web100_snap_read(var, snap, buf)) {
                    web100_perror("web100_snap_read");
                    exit(EXIT_FAILURE);
                }

                printf("%-20s %s\n",
                       web100_get_var_name(var),
                       web100_value_to_text(web100_get_var_type(var),
                                            buf));

                var = web100_var_next(var);
            }

            web100_snapshot_free(snap);

            conn = web100_connection_next(conn);

            if (conn)
                printf("\n");
        }

        group = web100_group_next(group);

        if (group)
            printf("\n");
    }

    return 0;
}
